display_list/geometry/dl_path.cc

```diff
@@ -9,64 +9,75 @@
 
 namespace flutter {
 
-const SkPath& DlPath::GetSkPath() const {
-  return sk_path_;
+const SkPath& DlPath::GetSkPath(bool will_render) const {
+  if (will_render) {
+    if (data_->use_count >= kMaxVolatileUses) {
+      data_->sk_path.setIsVolatile(false);
+    } else {
+      data_->use_count++;
+    }
+  }
+  return data_->sk_path;
 }
 
 impeller::Path DlPath::GetPath() const {
-  if (path_.IsEmpty() && !sk_path_.isEmpty()) {
-    path_ = ConvertToImpellerPath(sk_path_);
+  if (data_->path.IsEmpty() && !data_->sk_path.isEmpty()) {
+    data_->path = ConvertToImpellerPath(data_->sk_path);
   }
 
-  return path_;
+  return data_->path;
 }
 
 bool DlPath::IsInverseFillType() const {
-  return sk_path_.isInverseFillType();
+  return data_->sk_path.isInverseFillType();
 }
 
 bool DlPath::IsRect(DlRect* rect, bool* is_closed) const {
-  return sk_path_.isRect(ToSkRect(rect), is_closed);
+  return data_->sk_path.isRect(ToSkRect(rect), is_closed);
 }
 
 bool DlPath::IsOval(DlRect* bounds) const {
-  return sk_path_.isOval(ToSkRect(bounds));
+  return data_->sk_path.isOval(ToSkRect(bounds));
 }
 
 bool DlPath::IsSkRect(SkRect* rect, bool* is_closed) const {
-  return sk_path_.isRect(rect, is_closed);
+  return data_->sk_path.isRect(rect, is_closed);
 }
 
 bool DlPath::IsSkOval(SkRect* bounds) const {
-  return sk_path_.isOval(bounds);
+  return data_->sk_path.isOval(bounds);
 }
 
 bool DlPath::IsSkRRect(SkRRect* rrect) const {
-  return sk_path_.isRRect(rrect);
+  return data_->sk_path.isRRect(rrect);
 }
 
 SkRect DlPath::GetSkBounds() const {
-  return sk_path_.getBounds();
+  return data_->sk_path.getBounds();
 }
 
 DlRect DlPath::GetBounds() const {
-  return ToDlRect(sk_path_.getBounds());
+  return ToDlRect(data_->sk_path.getBounds());
 }
 
 bool DlPath::operator==(const DlPath& other) const {
-  return sk_path_ == other.sk_path_;
+  return data_->sk_path == other.data_->sk_path;
 }
 
 bool DlPath::IsConverted() const {
-  if (!path_.IsEmpty()) {
+  if (!data_->path.IsEmpty()) {
```

Comment:
I don't know if we should use something like isEmpty here. if we're given a totally empty, then we'd keep reconverting it each frame ... though I guess if it is an empty path maybe that doesn't matter.

I think the fact that this method works does mean that we're default initializing the internal impeller::Path shared_ptr, and then overwriting it later with a converted path. Maybe adding the optional would help with that.

Reply:
This just echoes the logic in the GetPath method. I think if the sk_path is empty we just leave the impeller::Path at its default.